src/cpp/src/visual_language/cdpruner/cdpruner.cpp

```diff
@@ -0,0 +1,377 @@
+    size_t actual_selected_tokens = selected_tokens.empty() ? 0 : selected_tokens[0].size();
+
+    ov::Tensor pruned_features(visual_features.get_element_type(), {batch_size, actual_selected_tokens, feature_dim});
+
+    const float* input_data = visual_features.data<const float>();
+    float* output_data = pruned_features.data<float>();
+
+    for (size_t b = 0; b < batch_size; ++b) {
+        const auto& batch_selected = selected_tokens[b];
+
+        for (size_t t = 0; t < batch_selected.size(); ++t) {
+            size_t src_token_idx = batch_selected[t];
+
+            for (size_t f = 0; f < feature_dim; ++f) {
+                size_t src_idx = b * total_tokens * feature_dim + src_token_idx * feature_dim + f;
+                size_t dst_idx = b * actual_selected_tokens * feature_dim + t * feature_dim + f;
+                output_data[dst_idx] = input_data[src_idx];
+            }
+        }
+    }
+
+    m_last_statistics.total_tokens = total_tokens;
+    m_last_statistics.selected_tokens = actual_selected_tokens;
+    m_last_statistics.pruning_ratio = 1.0f - static_cast<float>(actual_selected_tokens) / total_tokens;
+    m_last_statistics.batch_size = batch_size;
+
+    return pruned_features;
+}
+
+ov::Tensor CDPruner::apply_pruning(const std::vector<ov::Tensor>& visual_features_list,
+                                   const ov::Tensor& text_features) {
+    if (visual_features_list.empty()) {
+        return ov::Tensor();
+    }
+
+    // Handle single feature case by calling existing method
+    if (visual_features_list.size() == 1) {
+        return apply_pruning(visual_features_list[0], text_features);
+    }
+
+    const auto& first_feature = visual_features_list[0];
+    const auto& visual_shape = first_feature.get_shape();
+    const auto& text_shape = text_features.get_shape();
+    size_t batch_size = visual_shape[0];
+    size_t tokens_per_frame = visual_shape[1];
+    size_t feature_dim = visual_shape[2];
+    size_t total_input_tokens = tokens_per_frame * visual_features_list.size();
+
+    size_t num_tokens_to_keep =
+        static_cast<size_t>(std::round(tokens_per_frame * (1 - m_config.pruning_ratio / 100.0)));
+
+    // Apply pruning to each visual feature and collect results (using silent mode)
+    std::vector<ov::Tensor> pruned_features_list;
+    pruned_features_list.reserve(visual_features_list.size());
+    std::vector<std::vector<size_t>> aggregated_selected;
+
+    size_t global_offset = 0;
+
+    for (size_t frame_idx = 0; frame_idx < visual_features_list.size(); ++frame_idx) {
+        const auto& visual_feature = visual_features_list[frame_idx];
+        ov::Tensor pruned_feature = apply_pruning(visual_feature, text_features, true);
+        const auto& frame_selected = m_last_selected_tokens;
+        if (aggregated_selected.empty()) {
+            aggregated_selected.resize(frame_selected.size());
+        }
```

Comment:
`m_last_selected_tokens` is known variable before the for loop, please move this resize to before the for loop